### PR TITLE
drivers/ng_at86rf2xx: initial commit of refactoring

### PR DESCRIPTION
**Pull request by thomaseichinger**
CAUTION: This is not meant to be merged.

This is a prototype of the refactored driver for at86rf2xx drivers
intended to provide a basis to be enhanced for specific models and
implements the `ng_netdev_t` interface to use with ng_* network stack.

Will need special attention for the sending and receiving part and also 
to improve code capsulation for usage for specialised drivers.

I will have only limited time to work on this in the near future, so __feel free
to take it and improve/finish it__.

### CODE REVIEW

File: drivers/ng_at86rf2xx/ng_rf2xx.c [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+}
+
+void ng_at86rf2xx_on(ng_at86rf2xx_t *dev)
+{
+    /* Send a FORCE TRX OFF command */
+    ng_at86rf2xx_reg_write(dev, NG_AT86RF2XX_REG__TRX_STATE, NG_AT86RF2XX_TRX_STATE__FORCE_TRX_OFF);
+
+    /* busy wait for TRX_OFF state */
+    do {
+        int delay = _MAX_RETRIES;
```
Comment (LudwigOrtmann) [inline drivers/ng_at86rf2xx/ng_rf2xx.c]: compare https://github.com/RIOT-OS/RIOT/pull/2704

File: drivers/ng_at86rf2xx/ng_rf2xx.c [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+            DEBUG("at86rf2xx : ERROR : could not enter TRX_OFF mode\n");
+            return;
+        }
+    } while (ng_at86rf2xx_get_status(dev) != NG_AT86RF2XX_TRX_STATUS__TRX_OFF);
+
+    ng_at86rf2xx_reg_write(dev, NG_AT86RF2XX_REG__TRX_STATE, NG_AT86RF2XX_TRX_STATE__RX_ON);
+
+    /* change into basic reception mode to initialise CSMA seed by RNG */
+    do {
+        int delay = _MAX_RETRIES;
```
Comment (LudwigOrtmann) [inline drivers/ng_at86rf2xx/ng_rf2xx.c]: compare https://github.com/RIOT-OS/RIOT/pull/2704

File: drivers/ng_at86rf2xx/ng_rf2xx.c [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+    /*put pins to default values */
+    gpio_set(dev->cs_pin);
+    gpio_clear(dev->sleep_pin);
+    /* additional waiting to comply to min rst pulse width */
+    uint8_t volatile delay = 50; /* volatile to ensure it isn't optimized away */
+    while (--delay);
+    gpio_set(dev->reset_pin);
+}
+
+void ng_at86rf2xx_on(ng_at86rf2xx_t *dev)
```
Comment (LudwigOrtmann) [inline drivers/ng_at86rf2xx/ng_rf2xx.c]: re-add error handling (return error code)?

File: drivers/ng_at86rf2xx/ng_rf2xx.c [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    do {
+        int max_wait = _MAX_RETRIES;
+        if (!--max_wait) {
+            DEBUG("at86rf2xx : ERROR : could not enter RX_AACK_ON mode\n");
+            break;
+        }
+    } while (ng_at86rf2xx_get_status(dev) != NG_AT86RF2XX_TRX_STATUS__RX_AACK_ON);
+}
+
+void ng_at86rf2xx_initialise(ng_at86rf2xx_t *dev)
```
Comment (LudwigOrtmann) [inline drivers/ng_at86rf2xx/ng_rf2xx.c]: re-add error handling?

**Comment by haukepetersen** [comment]
@everyone: I will take the taken for this driver. I already started to refactor it and stumbled upon some design flaws. I hope I will be able to post an updated PR next week.

**Comment by bapclenet** [comment]
@haukepetersen, are you going to work on the implementation of 231, 233 and 212b within this new driver as well?
What will be required to implement to make each transceiver work with this new driver?

**Comment by haukepetersen** [comment]
@bapclenet I just started to look into the driver and basically refactored it to some more stable structure so far. Once I have it stable working (I plan to have this in the next days), I will also look into the variants other than the 231 -> here I will need help! So how about we do it this way: I PR my changes as soon as I have them ready and we have then a discussion on top of that code how we best extend it for the other variants?!

**Comment by bapclenet** [comment]
@haukepetersen, sounds great!
I will have a look at your PR when it's ready.

**Comment by haukepetersen** [comment]
I'll let you know :-)

**Comment by bapclenet** [comment]
@haukepetersen, Is [that](https://github.com/haukepetersen/RIOT/tree/opt_driver_rf2xx) your working branch?
Basically, once done, it should be ready to use the new Network Stack?

**Comment by haukepetersen** [comment]
I continued this work in #2825, so I take the freedom to close this PR.

**Pull request closed without merging**